feat(transactional-test-runner): Make so that a "simple" Abstract Account could be created more easily

Dkwcs opened the pull request:
[run-ci]

# Description of change
Added subcommand is `publish-deps`.

A new `publish-deps` subcommand that compiles and publishes support Move modules from a given path.
Path can be relative to iota root directory or absolute.

It looks like:

```
//# init --addresses test=0x0 --accounts A

//# publish-deps --paths crates/iota-adapter-transactional-tests/data/account_abstraction/abstract_account.move

//# publish --sender A --dependencies aa
```

## Links to any relevant issues

Fixes #9277 .

## How the change has been tested

Run `iota-adapter-transactional-tests` tests
With `cargo simtest`

Dkwcs committed:
feat(iota-types, iota-tx-checks, iota-protocol-config): new struct move_authenticator (#8138)

# Description of change

- Added `MoveAuthenticator` new variant for `GenericSignature` that
implements `AuthenticatorTrait` trait.
- new protocol variable `max_auth_gas` has been provided and set as
1000000000.
- openrpc and protocol snapshots have been updated.

## Links to any relevant issues

Fixes #8118 .

## How the change has been tested

Check snapshots - `cargo insta review`

Dkwcs committed:
feat(iota-types): authcontext, authdigeset new types have been created (#8179)

# Description of change

- Added new types that represent `AuthContext` and
`MoveAuthenticatorDigest`.
- AuthContext contains `MoveAuthenticatorDigest` and tx_inputs,
tx_commands(ProgrammableTransaction content)
Comparing to `TransactionDigest` only covers for the TransactionData
part,`MoveAuthenticatorDigest ` only covers for the `GenericSignature`
part and it is only needed during the `authenticate()` execution

## Links to any relevant issues

Fixes #8164 .

## How the change has been tested

Check snapshots - `cargo insta review`

valeriyr committed:
feat(iota-execution): add a `validate_transaction` function (#8131)

# Description of change

A new function has been added to `Executor` that implements smart
account transaction verification by calling the related Move
authenticator.

It is intended to be called during the processing of a smart account
transaction to unlock the corresponding smart account and allow the
transaction to be executed.

This function is integrated in the following PR:
https://github.com/iotaledger/iota/pull/8167

## Links to any relevant issues

fixes #8119

---------

Co-authored-by: Mirko Zichichi <miker83z@proton.me>

theiari committed:
feat: Added programmable_transaction.move and related test (#8146)

# Description of change

Introduced `programmable_transaction.move` and `auth_context.move` in
the framework, in order to represent these types in Move.

Closes #8122
Closes #8165 

## How the change has been tested

`programmable_transaction_tests` has been tested using `iota move test`

### Release Notes

---------

Co-authored-by: miker83z <mirko.zichichi@iota.org>

miker83z committed:
feat(framework-snapshot): create protocol v12 framework snapshots

valeriyr committed:
feat: validate a transaction with a `MoveAuthenticator` signature schema (#8167)

# Description of change

Integrated the `Executor:: validate_transaction` function implemented in
the following PR: https://github.com/iotaledger/iota/pull/8131

## Links to any relevant issues

fixes #8120

---------

Co-authored-by: Mirko Zichichi <miker83z@proton.me>

TheMrAI committed:
feat(iota-framework): Add create_auth_info_v1 native function (#8477)

# Description of change

Introduce create_auth_info_v1 native functions.
This function provides the capability of creating an AuthenticatorInfoV1
pointing to a desired authenticate
function.

The gas costs for the native function haven't been identified yet.
Tracking issue: #8433
The native function is not yet tested or integrated into the testing
frameworks. Tracking issue: #8416
The possibility of enabling the use of private `authenticate` function
will be investigated here: #8215

## Links to any relevant issues

fixes #8117
fixes #8216 

## How the change has been tested

The `create_auth_info_v1` feature hasn't been tested yet, but the
verifier was in a separate PR.

### Release Notes

- [x] Protocol: Introduce new move native function create_auth_info_v1
and create_auth_info_self_v1 with its gas costs.
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

Dkwcs committed:
feat(iota_account): implementation of a Move package to represent a basic abstracted IOTA account (#8250)

# Description of change

A new Move type has been introduced:

```
public struct IOTAccount has key {
  id: UID,
}
```

What can we do with the introduced account:
- Create an IOTAccount instance.
- Use account dynamic fields.
- Rotate the public key handled by the IOTAccount struct.
- Use basic authenticator functions.

## Links to any relevant issues

Fixes #8238 .

---------

Co-authored-by: Istvan Davorin Acs <istvan.acs@iota.org>
Co-authored-by: Valerii Reutov <valeriy.reutov@gmail.com>

TheMrAI committed:
fix(iota-framework): Update create_auth_info_v1 (#8532)

# Description of change

Now the AuthenticatorInfoV1 is constructed on the Move side, instead of
the Rust side.
This way we avoid the BSC conversion issue we had before. Native
function `create_auth_info_v1_impl` was also renamed to
`check_auth_info` as that is exactly what it does. Checks if everything
is correct, but it does not construct the AuthenticatorInfoV1.

## Links to any relevant issues

fixes #8528 

## How the change has been tested

Not tested.

### Release Notes

- [x] Protocol: `create_auth_info_v1_cost_base` changed to
`check_auth_info_v1_cost_base`
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

valeriyr committed:
fix: loading an authenticator info in `check_move_account` (#8553)

# Description of change

Fixed `AuthenticatorInfoV1` dynamic field ID generation when validating
an account object.

## Links to any relevant issues

fixes #8545

---------

Co-authored-by: Mirko Zichichi <miker83z@proton.me>
Co-authored-by: miker83z <mirko.zichichi@iota.org>

valeriyr committed:
fix: ptb arguments checking (#8574)

# Description of change

Fixed the following issue:
```
ErrorObject { code: ServerError(-32002), message: "Transaction execution failed due to issues with transaction inputs, please review the errors and try again:\n- Failed to execute the Move authenticator, reason: \"ExecutionError: ExecutionError { inner: ExecutionErrorInner { kind: CommandArgumentError { arg_idx: 2, kind: InvalidUsageOfPureArg }, source: Some(\\\"Non-primitive argument at index 2. If it is an object, it must be populated by an object\\\"), command: Some(0) } }\".", data: None }.
```

It is a temporary fix; the solution will be refactored to align with the
`TxContext` approach.

## Links to any relevant issues

fixes #8545

miker83z committed:
fix(auth_context): serialization uses TypeName instead of TypeInput (#8577)

# Description of change

Given that the `AuthContext` Move representation makes use of the
`TypeName`, i.e., the string representation of the type, its serializer
was implemented to support the conversion from `TypeInput` to
`TypeName`.

## Links to any relevant issues

Fixes #8455 issue 3

## How the change has been tested

- [ ] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

### Release Notes

- [ ] Protocol:
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

miker83z committed:
feat(docs): add an aa example using the rust sdk (#8543)

# Description of change

This example shows how to create an abstracted account and send a TX
through it using a Move-based Ed25519 authentication.

Here’s what it does step by step:
1. **Setup**
**IMPORTANT**: this example needs to run a local network with a faucet
manually (see instructions below for the exact command)
   * Connects to a local IOTA network using `IotaClientBuilder`.
* Initializes an in-memory keystore and imports a test account from a
fixed mnemonic.
   * Requests test tokens (faucet) for the sender.

3. **Create an IOTAccount**

* Publishes the **IOTAccount Move package** (which uses an Account
Abstraction logic) to the network.
* Uses the published package to create a new `IOTAccount` shared object.
   * Requests faucet tokens for the new account.

4. **Build a Simple Transaction**

* Prepares a sample PTB (`iota_account_simple_tx`) that calls
`iota_account::add_field` with dummy values.
   * Uses the `IOTAccount` as the sender.

5. **Sign & Authenticate**

* Signs the transaction data with the `IOTAccount` owner’s Ed25519 key.
* Encodes the signature and wraps it in a `MoveAuthenticator` so it can
be verified by the on-chain authenticator in Move.

6. **Execute Transaction**

* Submits the signed transaction with the Move-based authenticator to
the network.

## Links to any relevant issues

Fixes #8538

## How the change has been tested

- start a local network 
`cargo run --bin iota start --force-regenesis --with-faucet
--faucet-amount 100000000000000 --committee-size 2 --epoch-duration-ms
3600000`
- `cd docs/examples/rust`
- `cargo run --example iotaccount`
- The explorer can be helpful (runs only with some browser
configurations)
https://explorer.iota.org/?network=http%3A%2F%2F127.0.0.1%3A9000

valeriyr committed:
feat: first dynamic multisig account implementation (#8638)

# Description of change

This PR contains a `Dynamic Multisig Account` example implementation.

## Links to any relevant issues

fixes #8621

TheMrAI committed:
feat(iota-e2e-tests): Provide e2e tests for create_auth_info feature  (#8539)

# Description of change

Added a number of e2e tests for the
`iota-execution/latest/iota-verifier/src/account_auth_verifier.rs`.

## Links to any relevant issues

fixes #8414 
fixes #8525 

## How the change has been tested

- [ ] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

TheMrAI committed:
refactor(examples): Make iota_account reusable (#8744)

# Description of change

Refactor `iota_account` example to make it more reusable by extracting
its construction and protection aspects.
`iota_account` has its tests fully reworked, now they should cover 100%
all expected behavior.
`basic_keyed_account` has been left in the `iota_account` package for
now, for simplicity, but in the future we may extract it further. The
tests for this module have only been ported with no further
modifications. Some of it can/could be simplified, because with the
current setup certain scenarios are already forbidden, such as when
non-accounts attempt to modify state.

## Links to any relevant issues

fixes #8710 

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

valeriyr committed:
feat: add `AuthenticatorInfoV1Key` instead of `AUTHENTICATOR_DF_NAME` (#8759)

# Description of change

In `iota::account`, the string `AUTHENTICATOR_DF_NAME` is replaced with
a new type `AuthenticatorInfoV1Key`.

## Links to any relevant issues

fixes #8754

miker83z committed:
Merge branch 'vm-lang/aa-auth/8805-beta-feature-branch' into vm-lang/aa-auth/8116-feature-branch

miker83z committed:
Merge branch 'vm-lang/aa-auth/8805-beta-feature-branch' into vm-lang/aa-auth/8116-feature-branch

miker83z committed:
feat(e2e-tests): add abstract account create and issuing TX tests (#8738)

# Description of change

This PR adds a new set of e2e tests for the Account Abstraction feature:
- create an abstract account
- issue a TX using it
- issue a sponsored TX using it

This is based on an implementation of a Move module that supports basic
operations for an Abstract Account (it could be modified later).

This is also the base for a series of additional e2e tests in the
future.

## Links to any relevant issues

Fixes #8670 

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [x] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

miker83z committed:
chore(tx-checks): fix clippy

TheMrAI committed:
feat(examples): Add "Time locked account" account abstraction example (#8602)

# Description of change

Implement "Time locked account" example for abstract accounts.
This example explores the possibility of utilizing the authentication
logic.

Each authenticator type can have a number of "reserved" dynamic fields
and provide basic functions for attaching, detaching, rotating, checking
and borrowing that type.
Properly using these functions is the responsibility of the implementer,
but given the language restrictions this is a very flexible pattern.

## Links to any relevant issues

fixes #8534 

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [x] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

Dkwcs committed:
feat(examples/move): function key plugin implementation for IOTAccount (#8697)

# Description of change

This package adds plugin-based execution feature to an IOTAccount,
enabling a per-pubkey allow-set of callable functions identified as
`(package, module, function)`. The authenticate entrypoint implements a
dual flow:

- Owner flow: if the provided pubkey is the stored owner key, only
Ed25519 verification is required (no allow-set or command-count checks).
- User flow: otherwise, after Ed25519 verification, the PTB must contain
exactly one MoveCall, and that call must be present in the delegated
pubkey’s allow-set.

## Links to any relevant issues

Fixes #8680 .

## How the change has been tested

Run move tests for `function_key` package.
`iota move test`

valeriyr committed:
refactor: add a new `Validation` execution mode and refactor `AuthContext` creation (#8881)

# Description of change

Added a new Move execution mode named `Validation`. It is allowed to use
the `iota::auth_context::AuthContext` type only in this execution mode.
`iota::auth_context::AuthContext` is added to a transaction parameters
list in the same way as it is implemented for
`iota::tx_context::TxContext`

## Links to any relevant issues

fixes #8845

valeriyr committed:
feat: add the address check in MoveAuthenticator, allow only one(sender) MoveAuthenticator (#8963)

# Description of change

1. Improved `MoveAuthenticator:: verify_claims` implementation by
checking the author address.
2. Check a transaction's authenticator amount; allow only one
authenticator(sender) until we implement sponsor support.

## Links to any relevant issues

fixes #8747

miker83z committed:
feat(authority): support failing validation effects commit post-consensus (#8927)

# Description of change
This makes so that a move authentication passing pre-consensus
(validators sign the TX) and failing post-consensus produces some
effects (gas charging).

I think starting from `test_abstract_account_post_consensus_failure()`
is a good way to understand the logic behind the change.
Basically, what is needed, is to make it so that this can happen:
1) Create an AA TX and make the majority of validators to sign it; this
means that the Move authentication based on the AA shared object state
is CURRENTLY successful.
2) Then, tamper with the AA shared object state by altering the state of
the AA shared object (e.g., create and send for execution a second TX
that changes a field of the AA shared object).
3) Finally, submit for execution the original certificate, i.e., the one
obtained from the signature of the majority of validators. This PR
changes the behavior at this step:
- before this PR, the TX execution simply didn't go through but just
raised a Move authentication error within the iota-core;
- whilst, after this PR, the Move authentication error is treated as
execution error and thus producing effects; the result is that all input
objects are considered when creating the transaction effects and the gas
coin is reduced with the authentication computation gas cost execution.

This PR also merges input objects in the case of a successful execution.

In order to produce effects, the method
`authenticate_then_execute_transaction_to_effects()` was added to the
execution engine. This new method executes all the steps necessary to
create the TransactionEffects after the authentication execution. This
means that, after running `authenticate_transaction_inner()` and failing
(as well as succeeding), the execution continues with
`execute_transaction_to_effects_inner()` in order to finish the job,
i.e., creating effects for a failing (or successful) transaction.

## Links to any relevant issues

Fixes #8908 

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [x] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

valeriyr committed:
refactor: optimize authenticator inputs loading (#9047)

# Description of change

The changes made in this PR are aimed at optimizing the loading of
transaction inputs.

## Links to any relevant issues

fixes #8849

valeriyr committed:
refactor: avoid gas coins double check (#9100)

# Description of change

This PR contains a refactoring to avoid gas coins double-check during
transaction handling.

## Links to any relevant issues

fixes #8847

valeriyr committed:
feat: treat authenticator shared objects as TX input (#9132)

# Description of change

Authenticator shared objects are treated as TX input.

## Links to any relevant issues

fixes #8937
fixes #9131

miker83z committed:
feat(verifier): reject authenticate functions having a Receiving ref (#9118)

# Description of change

Changes the verifier check for the authenticate function in order to
reject any Receiving parameter.

## Links to any relevant issues

Fixes #9104 

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [x] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

miker83z committed:
feat(compiler): add a new known attribute based on Flavor and move the attr logic into iota_mode (#9187)

# Description of change

Added a generic FlavoredAttribute which can then allow us to define new
attributes within the iota_mode compiler part.

## Links to any relevant issues

Fixes #9186

## How the change has been tested

- [ ] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

miker83z committed:
feat(move-build, verifier): allow metadata in bytecode and fill metadata during package build (#9137)

# Description of change

This PR introduces a new protocol feature flag to allow the insertion of
metadata in the serialized compiled modules (bytecode). This was
operation was disabled before this PR, with the flag
`no_extraneous_module_bytes`.

This allows to fill this metadata field during the execution of
`build_from_resolution_graph`.

At the module deserializer level, the only performed check is that
metadata have only 1 field and the metadata key is hardcoded. During the
execution of the iota-verifier, the metadata value is checked for having
well formed metadata struct (later authenticate and view functions
checks will be performed here).

## Links to any relevant issues

Fixes #9135 

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [x] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

valeriyr committed:
feat: force push the `object_to_authenticate` as first parameter of the authenticator move call (#8986)

[run-ci]

# Description of change

`object_to_authenticate` is pushed as a Move authenticator call as the
first parameter.

## Links to any relevant issues

fixes #8902

valeriyr committed:
feat: implement a template of authenticator info checking during attaching with a proof (#9181)

[run-ci]

# Description of change

This PR contains two parts:
1. `AuthenticatorInfo` was extended with a generic parameter `Account`
which indicates the account type that can be unlocked with the related
authenticator function.
2. Introduced `AuthenticatorInfoV1CompatibilityProof` that represents a
proof that `AuthenticatorInfo` can be attached to an account. This can
be considered as a temporal solution since we have a better idea how to
proof `AuthenticatorInfo` and an account compatibility.
 
## Links to any relevant issues

fixes #8903

miker83z committed:
feat(compiler): add authenticator attribute (#9188)

# Description of change

To support defining authenticator functions for the IOTA abstract
account workflow the user must be able to
specify in Move code which functions should be considered
"authenticators".
Moreover in the future there may be more than one version of
authenticators so a version number must be supported as well.

This PR provides three formats to define a move authenticator with
version.
As a Named attribute:
```
#[authenticator]
public fun authenticator_fun(...)
```
In this case the version number is inferred to be 1.

As an Assigned attribute:
```
#[authenticator = 1]
public fun authenticator_fun(...)
```
Where the version number is assigned explicitly.

And as a Parametrized attribute:
```
#[authenticator(version = 1)]
public fun authenticator_fun(...)
```
In this format the user must explicitly spell out the version number and
assign a value to it. While the format is verbose, it may become a
preferred variant as it could potentially be extended with additional
features like:
```
#[authenticator(version = 1,  something_else = 53)]
```

## Implementation notes
Attributes do NOT reach the bytecode. Instead these are stored into the
`FunctionInfo` and this passed upward. Later PRs will deal with handling
the Authenticator attribute from within the `FunctionInfo` (in this PR,
just `iota_types::move_package::FnInfo` was extended to support the
authenticator version).

## Links to any relevant issues

Fixes #8860 

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [x] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

miker83z committed:
feat(move-build,verifier): fill authenticator metadata during package build and check its validity (#9198)

# Description of change

This PR builds on top of the previous ones in order to make so that the
`#[authenticator]` attribute is stored into the compiled module bytes.
This allows the verifier to retrieve the attribute and then check the
associated function for authentication validity.

## Links to any relevant issues

Fixes #9189

## How the change has been tested

- [ ] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

Dkwcs committed:
feat(iota-transactional-test-runner): Create the IotaSubcommand::AbstractTransaction (#9185)

# Description of change

A new `abstract` subcommand has been implemented, enabling
authentication of an abstract account via the transactional-runner.
This addition now allows testing the abstract account authentication
flow through the transactional-test-runner.
 
Test-runner documentation has been updated according to the new
subcommand.

## Links to any relevant issues

Fixes #9103 .

## How the change has been tested

Run `iota-adapter-transactional-tests`

valeriyr committed:
feat: add `MoveAuthenticator::validity_check` (#9306)

[run-ci]

# Description of change

`MoveAuthenticator::validity_check` implementation.

## Links to any relevant issues

fixes #9105

Dkwcs committed:
feat(iota-e2e-tests): e2e failing tests (#9052)

# Description of change

Implementing 2 e2e tests for specific aa authentication cases like:

Attempt to receive an object in authenticator.
Attempt to receive am object in abstract account itself.

## Links to any relevant issues

Fixes #9034 .

## How the change has been tested

Run e2e aa tests

---------

Co-authored-by: Mirko Zichichi <mirko.zichichi@iota.org>

valeriyr committed:
fix: check object to authenticate for duplication (#9358)

# Description of change

Include `object_to_authenticate` in `MoveAuthenticator::input_objects`
and `MoveAuthenticator::shared_objects` that fixes the arguments
duplication check as well.

Dkwcs committed:
refactor(iota-transactional-test-runner): handle the automatic generation of gas payment for the AbstractTransaction command
Remove unnecessary code as _public_key: vector<u8>

Dkwcs committed:
minor improvements

Dkwcs committed:
fix review comments, minor doc refactoring

valeriyr committed:
feat: use `SenderSignedData::input_objects` instead of `TransactionData::input_objects` (#9304)

# Description of change

Replace the function usage in several places, but there are some that
still need to be checked:
1. Indexing.
2. Replays.

## Links to any relevant issues

fixes #9278

valeriyr committed:
fix: move account check (#9431)

# Description of change

Fix `AuthenticatorInfoV1` reading during the move account check.

miker83z committed:
feat(execution): create a package metadata object at publish time (#9296)

# Description of change

This PR removes the previous method for verifying the authenticate
functions which worked at runtime. Then, it introduces the creation of a
Package Metadata immutable object during a package publishing/upgrade.
Finally, it allows to use such metadata to create `AuthenticatorInfoV1`
instances for accounts.

## Links to any relevant issues

Fixes #8861 
Fixes #8862 

## How the change has been tested

- [ ] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

Dkwcs committed:
feat(transactional-test-runner): init-abstract-acc subcommand has been added. It allows to reduce amount of ptb calls for creating abstract account

miker83z committed:
Merge branch 'vm-lang/aa-auth/8805-beta-feature-branch' into vm-lang/aa-auth/8116-feature-branch

miker83z committed:
fix(e2e): remove unnecessary tests (#9457)

# Description of change

Remove unnecessary e2e tests and update snapshots.

valeriyr committed:
feat: check authenticator inputs for consistency (#9451)

# Description of change

Check that authenticator input objects are consistent.

## Links to any relevant issues

fixes #9430

---------

Co-authored-by: Mirko Zichichi <mirko.zichichi@iota.org>

Dkwcs committed:
feat(transactional-test-runner): init-abstract-acc has 2 modes. Custom and default, with default mode test-runner uses precompiled abstract account package. Custom mode allows to set your own absract account module in-place

Dkwcs committed:
clippy fix

miker83z commented on crates/iota-transactional-test-runner/data/account_abstraction/abstract_account.move:
```diff
@@ -0,0 +1,42 @@
+    }
+
+
+    public fun receive_object(
```
This function should not be part of the default template. We can create a dedicated `AbstractAccountReceiving` in those tests having this function.

Dkwcs committed:
remove unnecessary clone

Dkwcs committed:
several_account creation tests

Dkwcs committed:
fix(move-transactional-test-runner): Make MaybeNamedCompiledModule keep the annotations (#9466)

# Description of change

This PR fixes the loss of function annotations in Move transactional
tests and ensures that authenticator information such (#[authenticator]
attributes) is correctly propagated into `CompiledModule` metadata
though `MaybeNamedCompiledModule`.
`MaybeNamedCompiledModule` now carries `function_infos` in order to fill
metadata during the package publishing

Iota-adapter-transactional abstract account tests have been fixed.

Added tests:
several_authenticators.move
without_attribute.move

## Links to any relevant issues

Fixes #9369 .

## How the change has been tested

Run `iota-adapter-transactional-tests` tests
With `cargo simtest`

Dkwcs committed:
A new iota-adapter-test-runner subcommand has been added, allowing independent publishing of dependencies for Move code used in tests. This helps avoid duplication and improves overall flexibility.

Dkwcs committed:
fmt fix

Dkwcs committed:
revert unnecessary changes

Dkwcs committed:
minor fix

Dkwcs committed:
fix clippy

Dkwcs committed:
minor improvements

Dkwcs committed:
Merge remote-tracking branch 'origin/vm-lang/aa-auth/8116-feature-branch' into vm-lang/aa-auth/9276-automatic-generation-of-gas-payment

Dkwcs committed:
Merge remote-tracking branch 'origin/vm-lang/aa-auth/9276-automatic-generation-of-gas-payment' into vm-lang/aa-auth/9277-abstract-account-creation-test-runner

valeriyr committed:
refactor: create dedicated native functions for the AA object creation (#9485)

[run-ci]

# Description of change

The account creation Move API was simplified by removing
`AuthenticatorInfoV1CompatibilityProof`.

## Links to any relevant issues

fixes #9459

miker83z commented on crates/iota-transactional-test-runner/src/args.rs:
```diff
@@ -321,6 +356,12 @@ impl<ExtraValueArgs: ParsableValue, ExtraRunArgs: Parser> clap::FromArgMatches
                 RunCommand::from_arg_matches(matches)?,
                 ExtraRunArgs::from_arg_matches(matches)?,
             ),
+            Some(("init-abstract-acc", matches)) => IotaSubcommand::InitAbstractAccount(
```
```suggestion
            Some(("init-abstract-account", matches)) => IotaSubcommand::InitAbstractAccount(
```

miker83z commented on crates/iota-transactional-test-runner/src/args.rs:
```diff
@@ -321,6 +356,12 @@ impl<ExtraValueArgs: ParsableValue, ExtraRunArgs: Parser> clap::FromArgMatches
+            Some(("init-abstract-acc", matches)) => IotaSubcommand::InitAbstractAccount(
+                InitAbstractAccountCommand::from_arg_matches(matches)?,
+            ),
+            Some(("publish-deps", matches)) => {
```
```suggestion
            Some(("publish-dependencies", matches)) => {
```

Dkwcs replied:
Done

miker83z commented on crates/iota-adapter-transactional-tests/data/account_abstraction/abstract_account.move:
```diff
@@ -0,0 +1,29 @@
+    let account_address = object::id_address(&account);
+    iota::transfer::share_object(account);
+    account_address
+}
```
```suggestion
public fun create(
    package_metadata: &PackageMetadata,
    module_name: ascii::String,
    function_name: ascii::String,
    ctx: &mut TxContext,
): address {
    let authenticator = account::create_auth_info_v1<AbstractAccount>(
        package_metadata,
        module_name,
        function_name,
    );
    let mut account = AbstractAccount { id: object::new(ctx) };
    let authenticator_compatibility_proof = account::check_auth_info_v1_compatibility(
        &account,
        authenticator,
    );
    account::attach_auth_info_v1(account.uid_mut(), authenticator_compatibility_proof);
    let account_address = object::id_address(&account);
    iota::transfer::share_object(account);
    account_address
}

public fun create_with_auth_info(
    authenticator: AuthenticatorInfoV1<AbstractAccount>,
    ctx: &mut TxContext,
): address {
    let mut account = AbstractAccount { id: object::new(ctx) };
    let authenticator_compatibility_proof = account::check_auth_info_v1_compatibility(
        &account,
        authenticator,
    );
    account::attach_auth_info_v1(account.uid_mut(), authenticator_compatibility_proof);
    let account_address = object::id_address(&account);
    iota::transfer::share_object(account);
    account_address
}
```

miker83z replied:
After this change, we can remove the newly added command `InitAbstractAccount` and just use this, similarly to the original but with just 1 command:
```
//# programmable --sender A --inputs object(3,1) "authenticate" "authenticate"
//> 0: aa::abstract_account::create(Input(0), Input(1), Input(2));
```

I suggest this because, now that we can use dependencies we don't need the complexity of adding a new command, but we can use the Move code of dependencies to make our life simpler.

Dkwcs replied:
I thought we had to follow the closest to the real example possible way in tests.
We could use this Move code that time.

miker83z replied:
The "real" example is hidden in tests, so better use it to our advantage like in this example above.

Dkwcs replied:
Done

miker83z commented on crates/iota-transactional-test-runner/src/test_adapter.rs:
```diff
@@ -1226,6 +1240,60 @@ impl MoveTestAdapter<'_> for IotaTestAdapter {
                 let output = self.object_summary_output(&summary, /* summarize */ false);
                 Ok(output)
             }
+            IotaSubcommand::InitAbstractAccount(InitAbstractAccountCommand {
```
I suggest to remove it, pls see  https://github.com/iotaledger/iota/pull/9444/files#r2603653726

Dkwcs committed:
fix review comments(remove init-abstract-command)

Dkwcs committed:
Merge remote-tracking branch 'origin/vm-lang/aa-auth/8116-feature-branch' into vm-lang/aa-auth/9276-automatic-generation-of-gas-payment

Dkwcs committed:
Merge remote-tracking branch 'origin/vm-lang/aa-auth/9276-automatic-generation-of-gas-payment' into vm-lang/aa-auth/9277-abstract-account-creation-test-runner

valeriyr committed:
feat: support `deny` for move authenticators (#9557)

# Description of change

Support `Deny` for Move authenticators on the signing phase.

## Links to any relevant issues

fixes #8856

Dkwcs committed:
docs update

Dkwcs committed:
fix review comments

Dkwcs committed:
Merge remote-tracking branch 'origin/vm-lang/aa-auth/8116-feature-branch' into vm-lang/aa-auth/9276-automatic-generation-of-gas-payment

Dkwcs committed:
fix fmt

Dkwcs committed:
deny snapshot fix

Dkwcs committed:
Merge remote-tracking branch 'origin/vm-lang/aa-auth/9276-automatic-generation-of-gas-payment' into vm-lang/aa-auth/9277-abstract-account-creation-test-runner

Dkwcs committed:
dprint fix

Dkwcs committed:
fix clippy

Dkwcs committed:
clippy fix

miker83z commented on crates/iota-transactional-test-runner/syntax.md:
```diff
@@ -1562,6 +1563,36 @@ Contents: iota::coin::Coin<iota::iota::IOTA> {
+
+The `publish-dependencies` subcommand (`PublishDeps` in Rust) is a lightweight helper for transactional tests that need to publish prepared dependency modules (e.g., Account Abstraction implementations) in order to avoid duplication/complex relations.
+It compiles and publishes Move source files provided by path and then stores the published modules in the adapter’s compiled state. As a result, the newly published package becomes available for subsequent commands and can be used as a proper dependency within built-in Move code.
+
```
"The package name of the dependency must be initialized to `0x0` with the init command, e.g., `aa=0x0`"

Dkwcs replied:
8ba99dc06c2388f6c36f76a2772deaa29d8d56ed

miker83z commented on crates/iota-adapter-transactional-tests/data/account_abstraction/abstract_account.move:
```diff
@@ -0,0 +1,45 @@
+    account_address
+}
+
+public fun uid_mut(self: &mut AbstractAccount): &mut UID {
```
Since we haven't find a usage for this, please remove it.

Dkwcs replied:
8ba99dc06c2388f6c36f76a2772deaa29d8d56ed

miker83z commented on crates/iota-adapter-transactional-tests/data/account_abstraction/abstract_account_with_pub_key.move:
```diff
@@ -0,0 +1,47 @@
+    account_address
+}
+
+public fun uid_mut(self: &mut AbstractAccount): &mut UID {
```
Since we haven't find a usage for this, please remove it

Dkwcs replied:
8ba99dc06c2388f6c36f76a2772deaa29d8d56ed

miker83z commented on crates/iota-adapter-transactional-tests/tests/abstract_account/account/receive_object.move:
```diff
@@ -30,8 +40,9 @@ public fun create(
 
-#[authenticator]
-public fun authenticate(_account: &AbstractAccount, _auth_ctx: &AuthContext, _ctx: &TxContext) {}
+public fun uid_mut(self: &mut AbstractAccount): &mut UID {
```
Since we haven't find a usage for this, please remove it

Dkwcs replied:
8ba99dc06c2388f6c36f76a2772deaa29d8d56ed

Dkwcs committed:
fix review comments

valeriyr commented on crates/iota-adapter-transactional-tests/tests/abstract_account/account/mutable.move:
```diff
@@ -3,39 +3,23 @@
-//# init --addresses test=0x0 --accounts A
+//# init --addresses test=0x0 aa=0x0 --accounts A
 
-//# publish --sender A
-module test::abstract_account;
+//# publish-dependencies --paths crates/iota-adapter-transactional-tests/data/account_abstraction/abstract_account.move
```
Does it make sense to name the package(address) the same as the file name for clarity?
If we have a test with several account types, they will have different addresses.

Dkwcs replied:
It can make sense actually, however I don't get your second point, how is it related to the first one?

valeriyr replied:
As I see, `abstract_account.move` and `abstract_account_with_pub_key.move` have the same address `aa`.
By using a different address we can have a test like this:
```
//# init --addresses abstract_account=0x0 abstract_account_with_pub_key=0x0 --accounts A

//# publish-dependencies --paths crates/iota-adapter-transactional-tests/data/account_abstraction/abstract_account.move
//# publish-dependencies --paths crates/iota-adapter-transactional-tests/data/account_abstraction/abstract_account_with_pub_key.move

//# publish --sender A --dependencies abstract_account abstract_account_with_pub_key

...
```

valeriyr replied:
We can use different utilities to create a test where the authenticator info is rotated:
```
//# init --addresses abstract_account=0x0 keyed=0x0 multisig=0x0 --accounts A

//# publish-dependencies --paths crates/iota-adapter-transactional-tests/data/account_abstraction/abstract_account.move
//# publish-dependencies --paths crates/iota-adapter-transactional-tests/data/account_abstraction/keyed.move
//# publish-dependencies --paths crates/iota-adapter-transactional-tests/data/account_abstraction/multisig.move

...
```

valeriyr replied:
So, the main point is to have different package names :)

Dkwcs replied:
I guess it makes sense
